File: src/lib/api/events.ts
// src/lib/api/events.ts
import api from "./axios";

// ============== تایپ‌ها ==============
export interface EventResponse {
  id: string;
  title: string;
  slug: string;
  description?: string;
  content?: string;
  cover_image?: string;
  start_date: string;
  end_date: string;
  duration?: string;
  capacity: number;
  price: number;
  location?: string;
  category?: string;
  is_active: boolean;
  is_featured: boolean;
  created_at: string;
  updated_at: string;
}

export interface EventCreate {
  title: string;
  slug: string;
  description?: string;
  content?: string;
  cover_image?: string;
  start_date?: string;
  end_date?: string;
  duration?: string;
  capacity: number;
  price: number;
  location?: string;
  category?: string;
  is_active?: boolean;
  is_featured?: boolean;
}

export interface EventUpdate {
  title?: string;
  slug?: string;
  description?: string;
  content?: string;
  cover_image?: string;
  start_date?: string;
  end_date?: string;
  duration?: string;
  capacity?: number;
  price?: number;
  location?: string;
  category?: string;
  is_active?: boolean;
  is_featured?: boolean;
}

export interface EventPaginatedResponse {
  total: number;
  page: number;
  size: number;
  items: EventResponse[];
}

// ============== تابع تولید اسلاگ ==============
export const generateSlug = (text: string): string => {
  if (!text || text.trim() === "") return "بدون-عنوان";

  return (
    text
      .trim()
      .toLowerCase()
      // ✅ حذف کاراکترهای خاص (فقط حروف فارسی، انگلیسی، اعداد و خط تیره باقی می‌مانند)
      .replace(/[^\u0600-\u06FFa-zA-Z0-9\s-]/g, "")
      // ✅ تبدیل فاصله به خط تیره
      .replace(/\s+/g, "-")
      // ✅ حذف خط تیره‌های تکراری
      .replace(/-+/g, "-")
      // ✅ حذف خط تیره از ابتدا و انتها
      .replace(/^-+|-+$/g, "") ||
    // ✅ اگر خالی شد، مقدار پیش‌فرض
    "بدون-عنوان"
  );
};

// ============== API ==============
export const eventsAPI = {
  // 📋 عمومی - دریافت لیست رویدادها
  getAll: async (params?: {
    page?: number;
    size?: number;
    search?: string;
    category?: string;
    is_active?: boolean;
  }): Promise<EventPaginatedResponse> => {
    const response = await api.get("/events", { params });
    return response.data;
  },

  // 📖 عمومی - دریافت با اسلاگ
  getBySlug: async (slug: string): Promise<EventResponse> => {
    const response = await api.get(`/events/slug/${slug}`);
    return response.data;
  },

  // 🔍 ادمین - دریافت با ID
  getById: async (id: string): Promise<EventResponse> => {
    const response = await api.get(`/events/${id}`);
    return response.data;
  },

  // ➕ ادمین - ایجاد رویداد جدید
  create: async (data: EventCreate): Promise<EventResponse> => {
    const response = await api.post("/events", data);
    return response.data;
  },

  // ✏️ ادمین - ویرایش رویداد
  update: async (id: string, data: EventUpdate): Promise<EventResponse> => {
    const response = await api.put(`/events/${id}`, data);
    return response.data;
  },

  // 🗑️ ادمین - حذف رویداد
  delete: async (id: string): Promise<{ message: string }> => {
    const response = await api.delete(`/events/${id}`);
    return response.data;
  },
};

File: app/schemas/events.py
# app/schemas/events.py
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime

class EventCreateSchema(BaseModel):
    title: str = Field(..., description="عنوان رویداد", example="همایش تازه‌های هوافضا")
    slug: str = Field(..., description="شناسه متنی یکتا جهت URL", example="aerospace-news-conference")
    description: Optional[str] = Field(None, description="توضیحات کوتاه", example="بررسی آخرین دستاوردهای صنعت هوافضا")
    content: Optional[str] = Field(None, description="متن کامل و جزئیات رویداد")
    cover_image: Optional[str] = Field(None, description="آدرس تصویر کاور", example="/uploads/events/cover.jpg")
    start_date: Optional[datetime] = Field(None, description="تاریخ و زمان شروع")
    end_date: Optional[datetime] = Field(None, description="تاریخ و زمان پایان")
    location: Optional[str] = Field(None, description="مکان برگزاری (در صورت حضوری)", example="سالن همایش‌های مرکزی")
    is_online: Optional[bool] = Field(False, description="برگزاری به صورت آنلاین")
    capacity: Optional[int] = Field(None, description="ظرفیت رویداد", example=100)
    price: Optional[float] = Field(0.0, description="قیمت ورودی (صفر برای رایگان)", example=50000.0)
    category: Optional[str] = Field(None, description="دسته‌بندی رویداد", example="کارگاه آموزشی")
    is_active: Optional[bool] = Field(True, description="وضعیت فعال/غیرفعال")

class EventUpdateSchema(BaseModel):
    title: Optional[str] = None
    slug: Optional[str] = None
    description: Optional[str] = None
    content: Optional[str] = None
    cover_image: Optional[str] = None
    start_date: Optional[datetime] = None
    end_date: Optional[datetime] = None
    location: Optional[str] = None
    is_online: Optional[bool] = None
    capacity: Optional[int] = None
    price: Optional[float] = None
    category: Optional[str] = None
    is_active: Optional[bool] = None

class EventResponseSchema(BaseModel):
    id: str
    title: str
    slug: str
    description: Optional[str] = None
    content: Optional[str] = None
    cover_image: Optional[str] = None
    start_date: Optional[datetime] = None
    end_date: Optional[datetime] = None
    location: Optional[str] = None
    is_online: bool
    capacity: Optional[int] = None
    price: float
    category: Optional[str] = None
    is_active: bool
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True

class EventPaginatedResponseSchema(BaseModel):
    total: int
    page: int
    size: int
    items: List[EventResponseSchema]